auslib/db.py

```diff
@@ -1880,14 +1890,15 @@ def hasPermission(self, username, thing, action, product=None, transaction=None)
         except ValueError:
             return False
 
-        # If a user has a permission that doesn't explicitly limit the type of
-        # actions they can perform, they are allowed to do any type of action.
-        if options.get("actions") and action not in options["actions"]:
-            return False
-        # Similarly, permissions without products specified grant that
-        # that permission without any limitation on the product.
-        if options.get("products") and product not in options["products"]:
-            return False
+        if options:
+            # If a user has a permission that doesn't explicitly limit the type of
+            # actions they can perform, they are allowed to do any type of action.
+            if options.get("actions") and action not in options["actions"]:
+                return False
+            # Similarly, permissions without products specified grant that
+            # that permission without any limitation on the product.
+            if options.get("products") and product not in options["products"]:
+                return False
 
         return True
```

Comment:
Nice simplication! Thanks to it, I believe we can extract the conditions into function with more business-oriented names:
```py
def hasPermission(self, username, thing, action, product=None, transaction=None):
    perm = self.getPermission(username, "admin", transaction=transaction)
    if perm:
        return self._isProductAllowed(perm)    # Admins don't have actions.

    perm = self.getPermission(username, thing, transaction=transaction)
    if perm:    
        return all([self._isProductAllowed(perm), self._isActionAllowed(perm)])

    return False

def _isProductAllowed(permissions, product):
    return self._isFieldAbsentOrExplicitlyWhitelisted(permissions, 'products', product)

def _isActionAllowed(permissions, action):
    return self._isFieldAbsentOrExplicitlyWhitelisted(permissions, 'actions', action)

def _isFieldAbsentOrExplicitlyWhitelisted(permissions, field, field_value):
    # If the field is missing, we assume that the user has access over all elements of this field.
    options = perm["options"]
    return not options or not options.get(field) or field_value in options[field]
```